### PR TITLE
[nn] mha : no-batch-dim support (python)

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: https://github.com/pytorch/pytorch/issues/60585

* [x] Update docs
* [x] Tests for shape checking

Tests take roughly 20s on system that I use. Below is the timings for slowest 20 tests.

```
pytest test/test_modules.py -k _multih --durations=20
============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.10.0, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /home/kshiteej/Pytorch/pytorch_no_batch_mha, configfile: pytest.ini
plugins: hypothesis-6.23.2, repeat-0.9.1
collected 372 items / 336 deselected / 36 selected                                                                                                                                                               

test/test_modules.py ..............ssssssss..............                                                                                                                                                  [100%]

================================================================================================ warnings summary ================================================================================================
../../.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/backends/cudnn/__init__.py:73
test/test_modules.py::TestModuleCUDA::test_factory_kwargs_nn_MultiheadAttention_cuda_float32
  /home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.10/site-packages/torch/backends/cudnn/__init__.py:73: UserWarning: PyTorch was compiled without cuDNN/MIOpen support. To use cuDNN/MIOpen, rebuild PyTorch making sure the library is visible to the build system.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================== slowest 20 durations ==============================================================================================
8.66s call     test/test_modules.py::TestModuleCUDA::test_gradgrad_nn_MultiheadAttention_cuda_float64
2.02s call     test/test_modules.py::TestModuleCPU::test_gradgrad_nn_MultiheadAttention_cpu_float64
1.89s call     test/test_modules.py::TestModuleCUDA::test_grad_nn_MultiheadAttention_cuda_float64
1.01s call     test/test_modules.py::TestModuleCUDA::test_factory_kwargs_nn_MultiheadAttention_cuda_float32
0.51s call     test/test_modules.py::TestModuleCPU::test_grad_nn_MultiheadAttention_cpu_float64
0.46s call     test/test_modules.py::TestModuleCUDA::test_forward_nn_MultiheadAttention_cuda_float32
0.45s call     test/test_modules.py::TestModuleCUDA::test_non_contiguous_tensors_nn_MultiheadAttention_cuda_float64
0.44s call     test/test_modules.py::TestModuleCUDA::test_non_contiguous_tensors_nn_MultiheadAttention_cuda_float32
0.21s call     test/test_modules.py::TestModuleCUDA::test_pickle_nn_MultiheadAttention_cuda_float64
0.21s call     test/test_modules.py::TestModuleCUDA::test_pickle_nn_MultiheadAttention_cuda_float32
0.18s call     test/test_modules.py::TestModuleCUDA::test_forward_nn_MultiheadAttention_cuda_float64
0.17s call     test/test_modules.py::TestModuleCPU::test_non_contiguous_tensors_nn_MultiheadAttention_cpu_float32
0.16s call     test/test_modules.py::TestModuleCPU::test_non_contiguous_tensors_nn_MultiheadAttention_cpu_float64
0.11s call     test/test_modules.py::TestModuleCUDA::test_factory_kwargs_nn_MultiheadAttention_cuda_float64
0.08s call     test/test_modules.py::TestModuleCPU::test_pickle_nn_MultiheadAttention_cpu_float32
0.08s call     test/test_modules.py::TestModuleCPU::test_pickle_nn_MultiheadAttention_cpu_float64
0.06s call     test/test_modules.py::TestModuleCUDA::test_repr_nn_MultiheadAttention_cuda_float64
0.06s call     test/test_modules.py::TestModuleCUDA::test_repr_nn_MultiheadAttention_cuda_float32
0.06s call     test/test_modules.py::TestModuleCPU::test_forward_nn_MultiheadAttention_cpu_float32
0.06s call     test/test_modules.py::TestModuleCPU::test_forward_nn_MultiheadAttention_cpu_float64
============================================================================================ short test summary info =============================================================================================
=========================================================================== 28 passed, 8 skipped, 336 deselected, 2 warnings in 19.71s ===========================================================================
```

cc @albanD @mruberry @jbschlosser @walterddr

### CODE REVIEW

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -376,6 +393,36 @@ def module_inputs_torch_nn_Embedding(module_info, device, dtype, requires_grad,
     ]
 
 
+def module_inputs_torch_nn_MultiheadAttention(module_info, device, dtype, requires_grad, **kwargs):
+    make_input = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+    samples = []
+    bool_vals = (True, False)
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:399]: Can you please add a comment before this section indicating that these generated samples are for validating no-batch-dim support?

In the future, we'll want to expand this function to include other samples as well (not for this PR).

File: torch/testing/_internal/common_modules.py [hunk 2]
```diff
@@ -247,6 +247,23 @@ def no_batch_dim_reference_criterion_fn(m, *args, **kwargs):
     return output
 
 
+def no_batch_dim_reference_mha(m, p, *args, **kwargs):
+    """Reference function for MultiheadAttention supporting no batch dimensions.
+
+    The module is passed the input and target in batched form with a single item.
+    The output is squeezed to compare with the no-batch input.
+    """
+    dim = 0 if kwargs.get('batch_first', True) else 1
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:256]: nit: `dim` -> `batch_dim` reads better imo

File: torch/testing/_internal/common_modules.py [hunk 3]
```diff
@@ -247,6 +247,23 @@ def no_batch_dim_reference_criterion_fn(m, *args, **kwargs):
     return output
 
 
+def no_batch_dim_reference_mha(m, p, *args, **kwargs):
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:250]: I wonder if there's some way to make this more generic? We'll need similar logic involving `batch_first` for the various RNNs.

Going down this route, I think we could make this fully generic if we had some reference function that was configurable in these ways:
* Accepts which inputs to squeeze / unsqueeze (both args and kwargs)
* Accepts which outputs to squeeze / unsqueeze
* Handles `batch_first` (for sequence-based modules e.g. MHA / RNNs)
* Handles `reduction` (for criterion modules)

wdyt about this?

File: torch/nn/modules/activation.py [hunk 4]
```diff
@@ -963,16 +963,20 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
     Args:
         query: Query embeddings of shape :math:`(L, N, E_q)` when ``batch_first=False`` or :math:`(N, L, E_q)`
             when ``batch_first=True``, where :math:`L` is the target sequence length, :math:`N` is the batch size,
-            and :math:`E_q` is the query embedding dimension ``embed_dim``. Queries are compared against
-            key-value pairs to produce the output. See "Attention Is All You Need" for more details.
+            and :math:`E_q` is the query embedding dimension ``embed_dim``. Queries can also be unbatched with shape
```
Comment (jbschlosser) [inline torch/nn/modules/activation.py:966]: Wording suggestion:
```
query: Query embeddings of shape :math:`(L, E_q)`, :math:`(L, N, E_q)` with ``batch_first=False`` or :math:`(N, L, E_q)`
       with ``batch_first=True``, where :math:`L` is the target sequence length, :math:`N` is the batch size,
       and :math:`E_q` is the query embedding dimension ``embed_dim``
```

wdyt about the idea of just including the unbatched option within the list of allowed shapes?

File: torch/nn/functional.py [hunk 5]
```diff
@@ -5114,6 +5114,48 @@ def multi_head_attention_forward(
...
+            assert attn_mask.dim() in (2, 3), \
+                ("For batched (3-D) `query`, expected `attn_mask` to be `None`, 2-D or 3-D"
+                 f" but found {attn_mask.dim()}-D tensor instead")
+    elif query.dim() == 2:
+        assert key.dim() == 2 and value.dim() == 2, \
+            ("For unbatched (2-D) `query`, expected `key` and `value` to be 2-D"
+             f" but found {key.dim()}-D and {value.dim()}-D tensors respectively")
+        query = query.unsqueeze(1)
+        key = key.unsqueeze(1)
+        value = value.unsqueeze(1)
```
Comment (jbschlosser) [inline torch/nn/functional.py:5137]: imo the unsqueezing should be done after the shape check logic with nice comments indicating what's going on in terms of squeeze -> do MHA logic -> unsqueeze

File: test/test_nn.py [hunk 6]
```diff
@@ -5283,6 +5283,64 @@ def test_multihead_attn_no_bias(self):
         self.assertIsNone(mha.in_proj_bias)
         self.assertIsNone(mha.out_proj.bias)
 
+    def test_multihead_attn_invalid_shape(self):
```
Comment (jbschlosser) [inline test/test_nn.py:5286]: nice coverage here!

File: test/test_modules.py [hunk 7]
```diff
@@ -359,8 +359,12 @@ def _can_be_noncontiguous(obj):
...
             with freeze_rng_state():
                 default_output = m(*input_args, **input_kwargs)
-                grad_output = default_output.clone().detach_().normal_()
-                default_output.backward(grad_output, retain_graph=True)
+                if isinstance(default_output, torch.Tensor):
+                    grad_output = default_output.clone().detach_().normal_()
+                    default_output.backward(grad_output, retain_graph=True)
+                else:
+                    grad_output = tuple(o.clone().detach_().normal_() for o in default_output)
+                    (o.backward(grad_output, retain_graph=True) for o in grad_output)
```
Comment (jbschlosser) [inline test/test_modules.py:367]: Looks like this logic was added because `test_non_contiguous_tensors` did not support multiple outputs - is that correct?

Is it possible to use `_traverse_obj()` defined above to simplify this?

File: test/test_modules.py [hunk 8]
```diff
@@ -380,7 +384,10 @@ def _can_be_noncontiguous(obj):
 
                 with freeze_rng_state():
                     out = m(*in_args, **in_kwargs)
-                    out.backward(g_out_copy, retain_graph=True)
+                    if isinstance(out, torch.Tensor):
+                        out.backward(g_out_copy, retain_graph=True)
+                    else:
+                        (o.backward(grad_output, retain_graph=True) for o in grad_output)
```
Comment (jbschlosser) [inline test/test_modules.py:390]: Similar question here regarding `_traverse_obj()`

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! There's a couple things we should think about to nail this down nicely before applying it to the other sequence-based modules. Nothing huge but I have some intuition it might be possible to make some things simpler through more generic reusable logic :)

File: torch/nn/modules/activation.py (hunk 4 above)
Comment (kshitij12345) [reply torch/nn/modules/activation.py]: Slightly tweaked that to 
```
:math:`(L, E_q)` for unbatched input, :math:`(L, N, E_q)` with ``batch_first=False`
```

because without that 
```
Query embeddings of shape :math:`(L, E_q)`, :math:`(L, N, E_q)` with ``batch_first=False``
```

seems like :math:`(L, E_q)` should be passed with `batch_first=False`.

File: test/test_modules.py (hunk 7 above)
Comment (kshitij12345) [reply test/test_modules.py]: Had to tweak this slightly as it was always returning `None`. Now it works correctly. I don't think `_traverse_obj` will work with this.

File: torch/testing/_internal/common_modules.py (hunk 3 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_modules.py:250]: I think MHA has a very unique structure
Eg. output having batch-dim at different position.
```
(output[0].squeeze(dim), output[1].squeeze(0))
```

I think instead of trying to write a complicated generic function we might keep it seperate for now. Going forward with RNNs and other Transformer related layer we can introduce the generic version.

File: torch/nn/modules/activation.py [hunk 9]
```diff
@@ -961,18 +961,21 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
...
     Args:
-        query: Query embeddings of shape :math:`(L, N, E_q)` when ``batch_first=False`` or :math:`(N, L, E_q)`
-            when ``batch_first=True``, where :math:`L` is the target sequence length, :math:`N` is the batch size,
-            and :math:`E_q` is the query embedding dimension ``embed_dim``. Queries are compared against
-            key-value pairs to produce the output. See "Attention Is All You Need" for more details.
+        query: Query embeddings of shape :math:`(L, E_q)` for unbatched input, :math:`(L, N, E_q)` when ``batch_first=False``
+            or :math:`(N, L, E_q)` when ``batch_first=True``, where :math:`L` is the target sequence length,
+            :math:`N` is the batch size, and :math:`E_q` is the query embedding dimension ``embed_dim``.
+            Queries are compared against key-value pairs to produce the output. See "Attention Is All You Need" for more details.
         key: Key embeddings of shape :math:`(S, N, E_k)` when ``batch_first=False`` or :math:`(N, S, E_k)` when
```
Comment (jbschlosser) [inline torch/nn/modules/activation.py:968]: Need similar phrasing changes as above for `key`

File: torch/nn/modules/activation.py [hunk 10]
```diff
@@ -961,18 +961,21 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
...
             ``batch_first=True``, where :math:`S` is the source sequence length, :math:`N` is the batch size, and
-            :math:`E_k` is the key embedding dimension ``kdim``. See "Attention Is All You Need" for more details.
+            :math:`E_k` is the key embedding dimension ``kdim``. For unbatched 2D `query`, shape should be :math:`(S, E_k)`.
+            See "Attention Is All You Need" for more details.
         value: Value embeddings of shape :math:`(S, N, E_v)` when ``batch_first=False`` or :math:`(N, S, E_v)` when
```
Comment (jbschlosser) [inline torch/nn/modules/activation.py:972]: Need similar phrasing changes as above for `value`

File: torch/nn/modules/activation.py [hunk 11]
```diff
@@ -990,12 +993,17 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
     Outputs:
         - **attn_output** - Attention outputs of shape :math:`(L, N, E)` when ``batch_first=False`` or
           :math:`(N, L, E)` when ``batch_first=True``, where :math:`L` is the target sequence length, :math:`N` is
-          the batch size, and :math:`E` is the embedding dimension ``embed_dim``.
+          the batch size, and :math:`E` is the embedding dimension ``embed_dim``. If inputs are unbatched then shape is
```
Comment (jbschlosser) [inline torch/nn/modules/activation.py:996]: Need similar phrasing changes as above for `attn_output`

File: torch/nn/modules/activation.py [hunk 12]
```diff
@@ -990,12 +993,17 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
...
+          :math:`(L, E)`.
         - **attn_output_weights** - Attention output weights of shape :math:`(N, L, S)`, where :math:`N` is the batch
           size, :math:`L` is the target sequence length, and :math:`S` is the source sequence length. Only returned
-          when ``need_weights=True``.
+          when ``need_weights=True``. If inputs are unbatched then shape is :math:`(L, S)`.
```
Comment (jbschlosser) [inline torch/nn/modules/activation.py:1000]: Need similar phrasing changes as above for `attn_output_weights`

File: torch/testing/_internal/common_modules.py (hunk 3 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_modules.py:250]: > I think MHA has a very unique structure
Eg. output having batch-dim at different position.

This is true and it does make it tricky. I'm good with leaving it in place for now, but I hope there will be a consolidation opportunity for the various RNN types.

**Review by jbschlosser — COMMENTED** [review]
Looking good! Just need some small doc changes for consistent unbatched phrasing throughout `query`, `key`, `value`, `attn_output`, `attn_output_weights`

**Review by jbschlosser — APPROVED** [review]
LGTM thanks!

**Comment by t10-13rocket** [comment]
LGTM
